diffusion/models/autoencoder.py

```diff
@@ -0,0 +1,613 @@
+    Based on code from https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/models/networks.py
+
+    Args:
+        input_channels (int): Number of input channels. Default: `3`.
+        num_filters (int): Number of filters in the first layer. Default: `64`.
+        num_layers (int): Number of layers in the discriminator. Default: `3`.
+        padding (int): How much padding to use in the conv layers. Default: `0`.
+    """
+
+    def __init__(self, input_channels: int = 3, num_filters: int = 64, num_layers: int = 3, padding: int = 0):
+        super().__init__()
+        self.input_channels = input_channels
+        self.num_filters = num_filters
+        self.num_layers = num_layers
+        self.padding = padding
+
+        self.blocks = nn.ModuleList()
+        input_conv = nn.Conv2d(self.input_channels, self.num_filters, kernel_size=4, stride=2, padding=self.padding)
+        nn.init.kaiming_normal_(input_conv.weight, nonlinearity='linear')
+        self.blocks.append(input_conv)
+
+        out_filters = self.num_filters
+        for n in range(1, self.num_layers):
+            in_filters = self.num_filters * 2**(n - 1)
+            out_filters = self.num_filters * min(2**n, 8)
+            conv = nn.Conv2d(in_filters, out_filters, kernel_size=4, stride=2, padding=self.padding, bias=False)
+            # Init these as if a linear layer follows them because batchnorm happens before leaky relu.
+            nn.init.kaiming_normal_(conv.weight, nonlinearity='linear')
+            norm = nn.BatchNorm2d(out_filters)
+            nonlinearity = nn.LeakyReLU(0.2, True)
+            self.blocks.extend([conv, norm, nonlinearity])
+        # Make the output layers
+        final_out_filters = self.num_filters * min(2**self.num_layers, 8)
+        conv = nn.Conv2d(out_filters, final_out_filters, kernel_size=4, stride=1, padding=self.padding, bias=False)
+        nn.init.kaiming_normal_(conv.weight, nonlinearity='linear')
+        norm = nn.BatchNorm2d(final_out_filters)
+        nonlinearity = nn.LeakyReLU(0.2, True)
+        self.blocks.extend([conv, norm, nonlinearity])
+        # Output layer
+        output_conv = nn.Conv2d(final_out_filters, 1, kernel_size=4, stride=1, padding=self.padding, bias=False)
+        nn.init.kaiming_normal_(output_conv.weight, nonlinearity='linear')
+        # Should init output layer such that outputs are generally within the linear region of a sigmoid.
+        output_conv.weight.data *= 0.1
+        self.blocks.append(output_conv)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward through the discriminator."""
+        for block in self.blocks:
+            x = block(x)
+        return x
+
+
+class AutoEncoderLoss(nn.Module):
+    """Loss function for training an autoencoder. Includes LPIPs and a discriminator.
+
+    Args:
+        input_key (str): Key for the input to the model. Default: `image`.
+        output_channels (int): Number of output channels. Default: `3`.
```

Comment:
I suppose it's because the discriminator is taking the output from the AE, but it will also take the input as well? Eh not an important change...

Reply:
Yes, it's because this value should be the autoencoder's `output_channels` value. Not sure which way would be less confusing here, although typically `output_channels = input_channels` anyways.

Reply:
Understandable

Reply:
How annoying would it be to set this to `ae_output_channels`?

Reply:
Done!